### PR TITLE
Provide dockerx setup for docker builds

### PR DESCRIPTION
**Pull request by sikehish**
This address #1064.
A shell script was created to handle building and pushing multi-architecture images using docker buildx. 

## Changes  
This PR introduces a shell script (`buildx_and_push.sh`) to automate the process of building and pushing Docker images for **multiple architectures** using `docker buildx`. It ensures that the Docker images are always built with the **latest version of the `sf-hamilton-ui` library** from PyPI. 

### Key Changes:
1. **Multi-architecture support**:
   - Builds images for multiple platforms (`linux/amd64`, `linux/arm64`).
   
2. **Automated version management**:
   - The script dynamically fetches the latest `sf-hamilton-ui` version from **PyPI**.
   - The fetched version is used to tag Docker images, ensuring that both the versioned tag and `latest` tag are always in sync with the current release.

3. **Builder instance setup**:
   - Ensures that **Docker Buildx** is enabled and uses a named builder instance (`hamilton-builder`).

4. **Image tagging and pushing**:
   - Tags both backend and frontend images with:
     - The latest `sf-hamilton-ui` version.
     - The `latest` tag.

## How I Tested This  
1. **Environment**:
   - Tested on a local machine with Docker Buildx enabled.

2. **Test Cases**:
   - Checked the output to confirm that the fetched version from PyPI was used for tagging the images.
   - Built and pushed the backend and frontend images successfully using:
     ```bash
     ./buildx_and_push.sh
     ```
   - Verified that both versioned and `latest` tags were pushed to the repository.

3. **Validation**:
   - Confirmed that the following tags were pushed:
     - `dagworks/ui-backend:<fetched_version>`  
     - `dagworks/ui-backend:latest`  
     - `dagworks/ui-frontend:<fetched_version>`  
     - `dagworks/ui-frontend:latest`  

## Notes  
- **Ensure Docker Buildx is installed** before running the script. The setup logic in the script will handle creating the builder instance if it is not already configured.
- **Why this change is needed**: This automation ensures consistency between the versioned tag and the `latest` tag. It also simplifies multi-architecture builds and makes them easier to manage across multiple platforms.

- [x] PR has an informative and human-readable title (this will be pulled into the release notes)
- [x] Changes are limited to a single goal (no scope creep)
- [ ] Code passed the pre-commit check & code is left cleaner/nicer than when first encountered.
- [ ] Any _change_ in functionality is tested
- [ ] New functions are documented (with a description, list of inputs, and expected output)
- [ ] Placeholder code is flagged / future TODOs are captured in comments
- [ ] Project documentation has been updated if adding/changing functionality.

### CODE REVIEW

**Review by skrawcz — COMMENTED** [review]
Looks great!

what would a github action look like to get this to run once a new `sf-hamilton-ui` version was published?

**Comment by sikehish** [comment]
> Looks great!
> 
> what would a github action look like to get this to run once a new `sf-hamilton-ui` version was published?
I'll be working on the action soon. But I'm currently building the images without the env variables/args. What do I do? Do I add multi arch builds to the docker compose file, or is there any other approach that I need to take?

**Comment by skrawcz** [comment]
> I'll be working on the action soon. But I'm currently building the images without the env variables/args. What do I do? Do I add multi arch builds to the docker compose file, or is there any other approach that I need to take?

Sorry I'm not following. The multi-arch builds would just use your script here right?

**Comment by sikehish** [comment]
> > I'll be working on the action soon. But I'm currently building the images without the env variables/args. What do I do? Do I add multi arch builds to the docker compose file, or is there any other approach that I need to take?
> 
> Sorry I'm not following. The multi-arch builds would just use your script here right?

Yes. Just wanted to ask if you if the script is satisfying the requirements.

**Comment by skrawcz** [comment]
> Yes. Just wanted to ask if you if the script is satisfying the requirements.

Script looks great thanks. I haven't run it yet to verify though; so I assume it works right now.

**Comment by sikehish** [comment]
> Looks great!
> 
> what would a github action look like to get this to run once a new `sf-hamilton-ui` version was published?



> Looks great!
> 
> what would a github action look like to get this to run once a new `sf-hamilton-ui` version was published?

I believe , we can set up a workflow that triggers on the new release of the library. We can poll(using cron job) for the latest version. Does this sound fine?

File: .github/workflows/hamilton-ui-build-and-push.yml [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        id: compare_versions
+        run: |
+          PREVIOUS_VERSION=$(cat version_cache/previous_version.txt 2>/dev/null || echo "none")
+          echo "Previous version: $PREVIOUS_VERSION"
+          if [[ "$latest_version" == "$PREVIOUS_VERSION" ]]; then
+            echo "No new version found. Exiting."
+            exit 0
+          else
+            echo "New version detected: $latest_version"
+          fi
```
Comment (skrawcz) [inline .github/workflows/hamilton-ui-build-and-push.yml:49]: you could probably instead check what's pushed to docker hub instead? i.e. what is the highest version value? That might be more robust?

Comment (sikehish) [reply .github/workflows/hamilton-ui-build-and-push.yml:49]: > you could probably instead check what's pushed to docker hub instead? i.e. what is the highest version value? That might be more robust? Thoughts?
Well, In that case, I might not be able to check if the previous version(for the previous build)  is the same as the current version, and then would not be able to early exit if they're the same. So, that is why I've used the cache action.
The latest version is being obtained from the sf-hamilton-ui package that's there on PyPI .

**Comment by sikehish** [comment]
Hi @skrawcz 
I've made the reuqested changes. Let me know if anything else needs to be changed. I've tested it on my local fork, and its working fine.

File: ui/buildx_and_push.sh [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+BACKEND_IMAGE="dagworks/ui-backend"
+
+# define common platforms/architectures
+PLATFORMS="linux/amd64,linux/arm64"
+
+cd "$(dirname "$0")" # cd into the directory where this script is present(i.e. ui)
+
+docker buildx build --platform $PLATFORMS \
+    -t $BACKEND_IMAGE:$VERSION -t $BACKEND_IMAGE:latest \
+    --push -f backend/Dockerfile.backend backend/
```
Comment (skrawcz) [inline ui/buildx_and_push.sh:63]:
```suggestion
    --push -f backend/Dockerfile.backend-prod backend/
```

File: ui/buildx_and_push.sh [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+docker buildx build --platform $PLATFORMS \
+    -t $FRONTEND_IMAGE:$VERSION -t $FRONTEND_IMAGE:latest \
+    --push -f frontend/Dockerfile.frontend frontend/ \
```
Comment (skrawcz) [inline ui/buildx_and_push.sh:67]:
```suggestion
    --push -f frontend/Dockerfile.frontend-prod frontend/ \
```

**Review by skrawcz — APPROVED** [review]
just tried it - seems to work - i just had to change two small things

**Comment by sikehish** [comment]
> just tried it - seems to work - i just had to change two small things

Oh Okay. Anything else that I need to work on?

**Comment by sikehish** [comment]
> @sikehish just need to fix up the whitespace pre-commit issues. Otherwise this is ready to merge. Thanks!
> 
> e.g. `pip install pre-commit` `pre-commit run --all-files` I think those are the commands

I have executed the above command. Does everything look good now?
Thank you for the opportunity:)) It was fun working on this issue.

**Pull request merged**